### ISSUE TITLE
`batches per epoch` feature

### ISSUE THREAD

**Issue by maksym-del** [allenai/allennlp#4012]
**Is your feature request related to a problem? Please describe.**
Since allennlp's data iterators are gone, a useful feature to specify the epoch size is gone as well.

**Describe the solution you'd like**
Possibility to specify this parameter somewhere.

**Comment by dirkgr** [comment]
All the `DatasetReader`s take a `max_instances` parameter that makes them read only the first `n` instances. Does that do what you want?

**Comment by maksym-del** [comment]
@dirkgr the idea is to still pass through all the instances in the dataset. We just take a chunk of instances and decide that the epoch ends after this many instances.

DataIterator did exactly that:
https://github.com/allenai/allennlp/blob/8dc3a5b45f5b6441ca0acbf225a8d259522f02e3/allennlp/data/iterators/data_iterator.py#L64 
https://github.com/allenai/allennlp/blob/8dc3a5b45f5b6441ca0acbf225a8d259522f02e3/allennlp/data/iterators/data_iterator.py#L209
https://github.com/allenai/allennlp/blob/8dc3a5b45f5b6441ca0acbf225a8d259522f02e3/allennlp/data/iterators/data_iterator.py#L165

**Comment by dirkgr** [comment]
Do you want to do this so that you can run validation more often, and maybe early-stop in the middle of an epoch?

**Comment by matt-gardner** [comment]
I think the primary use case is to have the notion of an "epoch" (and whatever learning rate updates that entails) be decoupled from the actual size of your data, perhaps because your dataset is very large.  I'm not totally sure how to accomplish this in the new paradigm.  If we want to avoid modifying the trainer to handle this, then we'd need some kind of sampler that retains state.

**Comment by dirkgr** [comment]
I would want to stick with the definition of "epoch" being one run through the data. But it makes sense to me that you might want to do certain things more frequently than once per epoch.

**Comment by matt-gardner** [comment]
That definition really doesn't make sense in some cases, particularly for language modeling.  See, e.g., section 3.1.2 of the T5 paper: https://arxiv.org/pdf/1910.10683.pdf.  We use "epoch" to determine when checkpoints and validation happen, which really doesn't have to be after one pass through the data.  Enforcing that those have to align is pretty restrictive, and unnecessarily so.

**Comment by maksym-del** [comment]
Meanwhile, am I right that the easiest way to get checkpointing and evaluation done every N iterations right now is to define custom ```Checkpointer``` to serialize the model and define custom ```BatchCallback``` that will call https://github.com/allenai/allennlp/blob/01e2ad83afd612d5ebbb5ef93d0004a4d6dfd9a1/allennlp/training/util.py#L325 depending on the batch number?

**Comment by matt-gardner** [comment]
For checkpointing, you don't need a custom checkpointer, you just have to call the existing checkpointer every N iterations in a batch callback.  For evaluation, you could have the callback just call [`trainer._validation_loss`](https://github.com/allenai/allennlp/blob/master/allennlp/training/trainer.py#L563).

Hmm, I guess this kind of solves the problem, except for the learning rate stuff, which you can also handle just with a different learning rate scheduler.  And it's a bit awkward when you get to the end of the dataset, as the interactions with our existing training loop will cause an extra checkpoint and evaluation in an awkward spot.  But it's functional as a workaround until we think of something better.

**Comment by maksym-del** [comment]
I tried to implement this feature as a ```BatchCallback```, but faced several difficulties. 
The major one is that these batch callbacks are called from inside ```train_epoch``` function while the check for early stopping is done outside of this function. It means that even if ```MetricTracker``` already knows that we are out of patience (I want ```patience``` to be defined as a number of _mini-epochs_) one has to wait till the end of the full epoch to actually terminate training. For this to work, I would need to subclass ```GDTrainer``` and override train_epoch. 
Secondly, logging and other logic get slightly duplicated.
I came up with a cleaner solution, but I am leaving this as user feedback.

So in order to get the feature, we can define a wrapper for the pytroch ```DataLoader``` as following:
```python
class MiniEpochDataLoader():
    def __init__(self, data_loader: DataLoader, batches_per_epoch: int):
        self._batches_per_epoch = batches_per_epoch
        self.loader_iter = iter(data_loader)

    def __len__(self):
        return self._batches_per_epoch

    def __iter__(self):
        for i in range(self._batches_per_epoch): # defines mini-epoch generator size
            try:
                yield next(self.loader_iter)
            except StopIteration: # loader_iter is exhausted
                self.loader_iter = iter(data_loader) # so lets refresh it
                yield next(self.loader_iter) # and yield back required instance
```

It follows pytorch's ```DataLoader``` usage pattern of having ```__len__``` method and returning new ```Iterable``` each time it is called. So the wrapper can be used as drop-in replacement for it:
```python
class GDTrainer(...):
  def __init__(...):
        ...
        if batches_per_epoch is not None:
            self.data_loader = MiniEpochDataLoader(data_loader, batches_per_epoch)
        else:
            self.data_loader = data_loader
        ...
```
This code successfully works for me for training. If you like the approach I could wrap it as a PR.
